UED04 "Dual-Armed Husky" addition

UltronDestroyer opened the pull request:
Will be squashing commits/cleaning up as this goes on. This initial un-edited commit is to make users aware of the bulk of the work and where it started from.

In this pull request, I've added these features:
- the ability to use 2 UR5 arms on a Husky 
- the environment variable addition of the flir_ptu, bumblebee, velodyne vlp 16, bulkhead, and robotiq grippers
- changing from parametrization of launch files to environment variables for ur5, and following this convention forward in code
- addition of the husky_dual_ur5_moveit_config ROS package which holds all configurations for the Husky

Note: this pull request will require more than the Husky package to be a completion of work, it also depends on changes in (will reference the issues when I make them)
- husky_simulator
- robotiq
- universal_robot

Future work includes the addition of the octomap pipeline for moveit for the Dual UR5 platform, and the ability to control the flir_ptu via moveit

@tonybaltovski @paulbovbel @ibaranov-cp  If you could find the time - please take time to look over these changes and propose modifications or edits to the PR in case of convention.

TheClearpathDash committed:
Changed launch file to think of dual ur5

TheClearpathDash committed:
Added left and right arm for dual config, no positions yet, no top plate mod yet

TheClearpathDash committed:
Added robotiq grippers to launch file

TheClearpathDash committed:
Added robotiq grippers to dual husky

TheClearpathDash committed:
Added bulkhead as param

TheClearpathDash committed:
Added the joints for the bulkheead to arm

TheClearpathDash committed:
Changed default value to false

TheClearpathDash committed:
fixed slash syntax

TheClearpathDash committed:
Fixed xacro if

TheClearpathDash committed:
Almost compiles

TheClearpathDash committed:
Added option to include user rails or not

TheClearpathDash committed:
Added dual arm positioning

TheClearpathDash committed:
Added bulkhead positioning

TheClearpathDash committed:
Right arm positioned correctly

TheClearpathDash committed:
Positioned left arm for gazebo

TheClearpathDash committed:
Added ros controller files for gazebo for dual arm husky

TheClearpathDash committed:
Added config script for homing of dual ur5 in gazebo

TheClearpathDash committed:
set config script works

TheClearpathDash committed:
updated

TheClearpathDash committed:
Added stow scripts for left, right

TheClearpathDash committed:
Updated scripts

TheClearpathDash committed:
Updated cmake for stow

TheClearpathDash committed:
Updated scripts

TheClearpathDash committed:
Updated

TheClearpathDash committed:
added timeout

TheClearpathDash committed:
positioned grippers

TheClearpathDash committed:
Changed to use a different hand with a gazebo plugin

TheClearpathDash committed:
Updated position of grippers

TheClearpathDash committed:
Updated grpiper position for ft

TheClearpathDash committed:
Added description

TheClearpathDash committed:
fixed typog

TheClearpathDash committed:
updatd

TheClearpathDash committed:
upd

TheClearpathDash committed:
update ptu pos

TheClearpathDash committed:
positioned ptu

TheClearpathDash committed:
Updated

TheClearpathDash committed:
Added bumblebee2 mesh

TheClearpathDash committed:
Added ft300

TheClearpathDash committed:
Added dnyamic naming for laser pos

TheClearpathDash committed:
Added laser

TheClearpathDash committed:
Laser positioned and added

TheClearpathDash committed:
added camera

TheClearpathDash committed:
Positioned camera

TheClearpathDash committed:
Positoined bb

TheClearpathDash committed:
upd

TheClearpathDash committed:
Camera now upright

TheClearpathDash committed:
Positioned camera so its not tilted

TheClearpathDash committed:
Added new package with corrected urdf

TheClearpathDash committed:
Added ur5 kinematics

TheClearpathDash committed:
Added controllers.yaml

TheClearpathDash committed:
Changed name so it can find arm prefix

TheClearpathDash committed:
Changd name

TheClearpathDash committed:
Changed name

TheClearpathDash committed:
Added parent links

TheClearpathDash committed:
upd

TheClearpathDash committed:
Chnaged moveit pakg

TheClearpathDash committed:
Regend

TheClearpathDash committed:
Added passive joints

TheClearpathDash committed:
Added more bkg

TheClearpathDash committed:
upd

TheClearpathDash committed:
Controllers added

TheClearpathDash committed:
Defaulted some planner configs

TheClearpathDash committed:
Added limited parameter

TheClearpathDash committed:
Added limited to mg

TheClearpathDash committed:
Added limited to pc

TheClearpathDash committed:
ADded collision so it works w/ default

TheClearpathDash committed:
Orb for ptu works

TheClearpathDash committed:
Added param for rviz

TheClearpathDash committed:
upd

TheClearpathDash committed:
Updated

TheClearpathDash committed:
Finally connected sim controllers.. ffs

TheClearpathDash committed:
upd

TheClearpathDash committed:
removed unnec joint

TheClearpathDash committed:
update

TheClearpathDash committed:
Added home position for husky arms

TheClearpathDash committed:
Removed colllision state in home pose

TheClearpathDash committed:
Updated default interactive marker size

TheClearpathDash committed:
Added ft back

Commit:
Changed to optenv

TheClearpathDash committed:
Added dep on flir ptu

TheClearpathDash committed:
Merge branch 'dual_ur5_husky' of https://github.com/husky/husky into dual_ur5_husky

TheClearpathDash committed:
Added poitngrey camera description

TheClearpathDash committed:
Changed param

Commit:
Added optenvs

Commit:
Merge branch 'dual_ur5_husky' of https://github.com/husky/husky into dual_ur5_husky

Commit:
updated to optenv

Commit:
Added param for real robot

Commit:
Added remap

Commit:
Added 2 sep contrller files

Commit:
Added if else

Commit:
Added praram

Commit:
fix syntax

Commit:
fix syntax

Commit:
fix synt

Commit:
propogated arg

Commit:
add sim

TheClearpathDash committed:
positioned bumblebee

TheClearpathDash committed:
Merge branch 'dual_ur5_husky' of https://github.com/husky/husky into dual_ur5_husky

TheClearpathDash committed:
Added spacing for grippers

TheClearpathDash committed:
upd

TheClearpathDash committed:
fixed position of ptu

Commit:
update

Commit:
Update

Commit:
merged

Commit:
added jsc for ptu

Commit:
Comment out ptu stuff for now

Commit:
Add optenv for laser enabled

UltronDestroyer committed:
Update README.md

UltronDestroyer committed:
Update README.md

UltronDestroyer committed:
Update README.md

UltronDestroyer committed:
Update README.md

UltronDestroyer committed:
Update README.md

UltronDestroyer commented on husky_description/urdf/decorations.urdf.xacro:
```diff
@@ -43,21 +43,23 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
-      <parent link="base_link" />
-      <child link="user_rail_link" />
-    </joint>
+    <xacro:if value="$(optenv HUSKY_USER_RAILS true)">
```
@tonybaltovski @paulbovbel 

Note this will have a pretty big affect I imagine.

UltronDestroyer committed:
Update description.xacro.save

To be deleted.

UltronDestroyer commented on husky_description/urdf/husky.urdf.xacro:
```diff
@@ -141,14 +218,14 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
         <origin xyz="$(optenv HUSKY_LMS1XX_XYZ 0.2206 0.0 0.00635)"
                 rpy="$(optenv HUSKY_LMS1XX_RPY 0.0 0.0 0.0)" />
-        <parent link="top_plate_link" />
+        <parent link="$(optenv HUSKY_LASER_PARENT_LINK top_plate_link)" />
```
@tonybaltovski @paulbovbel 

With the addition of the bulk-head, I figured I needed to be able to easily change the parent link of the laser due to the structure that the payloads sit on changing all the time

paulbovbel replied:
The _XYZ and _RPY args are supposed to be relative to a common reference point in the solidworks model. Not a huge deal, although would be generally wary of adding env vars.

UltronDestroyer replied:
I once was opposed to using env vars. But now I see them as a great way to configure our model. I see no issue with having 100s of env vars and supporting a multitude of configurations.

Thoughts?

UltronDestroyer commented on husky_dual_ur5_moveit_config/launch/husky_dual_ur5_moveit_controller_manager.launch:
```diff
@@ -0,0 +1,18 @@
+      <rosparam file="$(find husky_dual_ur5_moveit_config)/config/real_controllers.yaml"/>
+  </group>
+
+  <group unless="$(arg real)">
```
@tonybaltovski @paulbovbel 

Simulation controllers are brought up unless the real robot is specified. This is because the real robot uses a different namespace for its controller/command scheme and the yaml files reflect that.

UltronDestroyer commented on husky_dual_ur5_moveit_config/launch/moveit_rviz.launch:
```diff
@@ -0,0 +1,17 @@
+  <node name="$(anon rviz)" launch-prefix="$(arg launch_prefix)" pkg="rviz" type="rviz" respawn="false"
+	args="$(arg command_args)" output="screen">
+    <rosparam command="load" file="$(find husky_dual_ur5_moveit_config)/config/kinematics.yaml"/>
+    <param name="arm_prefix" value="l_ur5_arm_"/>
```
I changed the way the universal_robot kinematics plugin acts, so this is unnecessary @TheDash TODO

@ibaranov-cp https://github.com/ros-industrial/universal_robot/pull/248/files

paulbovbel commented:
Squash your commits please

paulbovbel commented:
My perspective on the husky repositories is that it should include support for 'standard' configurations, and everything else belongs in a customer-specific customization repo. I would suspect that this Husky configuration is a fairly custom, oneoff piece of work. Please correct me if I'm wrong.

There is a lot of stuff here that makes sense for a one-off deployment...
- copy pasted yamls and python scripts for stowing/controller config, with minor changes - DRY, parametrize
- a gajillion new env vars that are related only to this dual-arm+bulkhead deployment and not general purpose
- namespace incompatibilities between sim and real robot, requiring a `real` roslaunch arg

.. but is a little worrying to put into the core husky packages

If this is going to be a standard product, let's talk more and we can do a more thorough break down and rework of the issues.

UltronDestroyer commented:
> If this is going to be a standard product, let's talk more and we can do a more thorough break down and rework of the issues.

Yes - this is a standard configuration for 2 armed Husky's. 

> My perspective on the husky repositories is that it should include support for 'standard' configurations, and everything else belongs in a customer-specific customization repo. I would suspect that this Husky configuration is a fairly custom, oneoff piece of work. Please correct me if I'm wrong.

We need to move away from this mentality. This repo should be configured via parameters that easily allow us to build a husky via building blocks like drivers, models, configurations. Similar to how the kernel can be configured for many use cases. This saves time in the future, and adds a process of progression and allows us to sell more options with less expense in engineering.

After a lot of drivers/setups configurations have been added, we can look to refactor that process to then strap onto other robots, so other robots we make can benefit from this process and use the drivers/system setup/system configuration we've made here.

A standard configuration is enabled by default, however, as always. 

> copy pasted yamls and python scripts for stowing/controller config, with minor changes - DRY, parametrize

I'll pass in a parameter "left" or "right" and use the same script. Can you comment on the code where there is copy and pasted yamls and offer a suggestion to "parametrize" and "refactor" a YAML file?

> a gajillion new env vars that are related only to this dual-arm+bulkhead deployment and not general purpose

... Not true. Later we can add things like env vars for positioning new sensors and their parents as well. See how the lasers are setup. I fail to see what's wrong with providing a 2 armed static setup. We don't currently have one. Are you suggesting I write a general purpose robot configuration setup for the Husky? If so, this PR moves closer to that goal by slowly adding more drivers/configurations such as the linux kernel does.

> namespace incompatibilities between sim and real robot, requiring a real roslaunch arg

This is a MoveIt! tradition and I want to preserve it.

TODOS:
1. Squash commits
2. Refactor stow scripts

mikepurvis commented:
I prefer to see envvars constrained to very simply flag-flips, like "enable the thing". Sophisticated configurations should use the xacro interface in a fully custom URDF, IMO.

I don't have enough context to comment on this in particular.

UltronDestroyer commented:
> I prefer to see envvars constrained to very simply flag-flips, like "enable the thing". Sophisticated configurations should use the xacro interface in a fully custom URDF, IMO.

This PR is set up so that there is an envvar to enable the dual arm basic setup, and another envvar to enable the grippers if they wish to use them.Without those two flags used, the husky spawns as standard. 
With the envvar which specifies only a single ur5, the single ur5 basic setup is loaded. 

At the moment, we have 3 configurations: a base model husky, a single armed husky, and a dual armed husky. Flipping on the respective env-vars handles everything behind the scene for each of these robots and so users should experience a regular launch.

UltronDestroyer commented:
Referencing the newly added parameters: http://wiki.ros.org/husky_bringup/Tutorials/Customize%20Husky%20Configuration#preview

paulbovbel commented:
Alright fair enough (regarding the standard product part), as long as it's done in a clean way. This and kernel build config is apples and firetrucks.

To justify my POV though, just remember that anything that goes into these repos will require a much heavier level of support than one-off config shipped to customers as customization, including but not limited to customers strapping their own bumblebees/etc to robots and expecting this to work for real robots and simulation.

Making something 'standard', and including it in this one repo for public consumption, is really only tenable when you're shipping >X of those configurations per year. I do lack context on what X is, so continuing on:
- Preference would be to make the parent_link for any non-dual-ur5 specific sensors (bumblebee, flir_ptu) to top_plate_link, so they can be used independently.
- Move any readme information to a section on the wiki, perhaps with specific pages for single and multi armed ur5
- Add the necessary dependencies to package.xml
- The control_*ur5.yaml's are pretty redundant, maybe template them using CMake or roslaunch XML's rosparam substitution mechanism.
- use the same stow script for single and two arm'd ur5 config

paulbovbel commented:
Also please squash :)

UltronDestroyer commented:
TODOS: (postponing merge until more time allocated)
1. Squash commits
2. Make parent link env vars for bumblebee, flir_ptu, velodyne.
3. move readme to tutorial for dual husky
4. add dependencies to package
5. Same stow script, parametrized
6. roslaunch param substitution? for the yaml files

wxmerkt commented:
@TheDash Is there any update regarding the referenced update to husky_simulator for the dual arm husky? I couldn't find a branch in the [corresponding repository](https://github.com/husky/husky_simulator/branches).

wxmerkt committed:
Generate URDF (old was out of date)

wxmerkt committed:
Delete auto-saved file

wxmerkt committed:
Fix PTU to Bulkhead joint z-offset (read from CAD model)

wxmerkt committed:
Fix Robotiq hand attachment z-offset (read from palm CAD model)

wxmerkt committed:
Add generate_dual_arm_husky_urdf.sh script as description.xacro won't work

wxmerkt committed:
Update URDF

wxmerkt committed:
Delete committed URDF, force users to generate from xacro

wxmerkt committed:
Update inertial link based on information provided by Clearpath support

wxmerkt committed:
Auto-point to dual-arm-husky xacro

wxmerkt committed:
Odd change required for catkin tools compatibility

wxmerkt committed:
Fix gripper mounting orientation approximately
Old orientation was off, calibrated visualisation vs real robot approximately to match

wxmerkt committed:
Change wheel multiplier to account for gear box

UltronDestroyer committed:
Merge pull request #36 from wxmerkt/dual_ur5_husky

Correct offsets in dual arm husky URDF

tonybaltovski commented on husky_control/config/control.yaml:
```diff
@@ -23,7 +23,7 @@ husky_velocity_controller:
   # Wheel separation and radius multipliers
   wheel_separation_multiplier: 1.875 # default: 1.0
-  wheel_radius_multiplier    : 1.0 # default: 1.0
+  wheel_radius_multiplier    : 0.5 # default: 1.0
```
This is not the default value.

UltronDestroyer replied:
That's correct. But it is required to be 0.5 for the gear box to work with the dual_ur5_husky setup. Thus, before this PR can be merged we need to change the control.yaml include so that it is different between setups and parametrized in the launch file.

tonybaltovski commented on husky_description/meshes/accessories/bumblebee2.dae:
```diff
@@ -0,0 +1,297 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Why not use the mesh from the driver: https://github.com/ros-drivers/pointgrey_camera_driver/tree/master/pointgrey_camera_description

UltronDestroyer replied:
Not sure why. But probably the better idea here.

UltronDestroyer replied:
removed the asset https://github.com/husky/husky/pull/33/commits/6ebc355d5ebda7a260aa809f494881b88d1e156e

TheClearpathDash committed:
Removed bumblebee2.dae mesh. No change to urdf.xacro, because xacro was already using pointgrey driver dae

tonybaltovski commented on husky_control/config/control_flir_ptu.yaml:
```diff
@@ -0,0 +1,15 @@
+  type: position_controllers/JointPositionController
+  joint: husky_ptu_pan
+  pid:
+    p: 0
```
Should these gains be all zero?

UltronDestroyer replied:
Not sure. Unable to test without a unit here.

tonybaltovski commented on husky_control/scripts/stow_ur5_right:
```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env python
```
Can the stow scripts be merged into one?

UltronDestroyer replied:
Yes, that can be done.

tonybaltovski commented on husky_description/urdf/description.xacro:
```diff
@@ -25,11 +25,13 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
-  <xacro:include filename="$(find husky_description)/urdf/husky.urdf.xacro" />
+  <xacro:arg name="robotiq_grippers_enabled" default="$(optenv ROBOTIQ_GRIPPERS_ENABLED false)" />
+ 
+  <xacro:include filename="$(find husky_description)/urdf/dual_arm_husky.urdf.xacro" />
```
Isn't this pointing to the wrong file?

UltronDestroyer replied:
Yes it is. the stuff inside dual_arm_husky.urdf.xacro needs to be merged with husky.urdf.xacro

tonybaltovski commented on husky_description/urdf/generate_dual_arm_husky_urdf.sh:
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
```
Is this script needed?  If so, can it be generic for any husky urdf?

UltronDestroyer replied:
It's not needed - but it's a nice to have. It can be generic, but I won't code it until someone needs it.

wxmerkt committed:
Fix IMU rotation

wxmerkt committed:
Fix bumblebee2 mount joint z-offset

wxmerkt committed:
Fix husky_ptu_joint x and y translation.
The y translation was taken from the FLIR PTU E46 data sheet and also confirmed via self-observation point clouds.
The x translation is guesstimated based on photographs and point cloud self-observation since I'm currently not in the lab. It is much better now and can be seen from the screw locations for the SICK mount for instance.

UltronDestroyer committed:
Merge pull request #39 from wxmerkt/dual_ur5_husky

Dual UR5 URDF Fixes: IMU rotation, PTU mount

wxmerkt commented on husky_control/config/control_left_ur5.yaml:
```diff
@@ -0,0 +1,22 @@
+l_arm_controller:
```
@TheDash these don't show up in the ``rosservice call /controller_manager/list_controllers`` list - might be related to #40

UltronDestroyer replied:
On the real robot, the intent of this is to be used for moveit's controller manager

paulbovbel commented:
In order to cut down on the amount of dependencies, everything UR/moveit-related has been removed from the base husky repos in kinetic-devel. Please see kinetic's `urdf_extras` and `config_extras` launch arguments as a hook to extend the base husky robot/simulation.

UltronDestroyer commented:
The dual UR5 armed husky org is here:

https://github.com/DualUR5Husky

UltronDestroyer commented:
Extensions are great - but modifications to the "base" systems aren't supported by kinetic, and so we've gone this route.

paulbovbel commented:
They are supported, via the `urdf_extras` and `config_extras` launch arguments and respective env vars.

UltronDestroyer commented:
They are not @paulbovbel. Launch arguments and env vars is way too complicated for development and actually slows the development time of getting a new robot out the door if there's new stuff to add.

If this stuff were useful, or perhaps it's not explained properly, I would use it. That being said - these do serve their purpose for the base husky customizations. As far as making a Husky more complex - this solution does not work.

Code complexity and spaghetti code is a real thing. And I'm avoiding it at all costs by not using the above soln for my problems
